### PR TITLE
Switch from `useCMEditViewDataManager()` to `useDocument()`

### PR DESCRIPTION
**Pull request by jackbuehner**
This PR finishes replacing usage of @strapi/helper-plugin. ([@strapi/helper-plugin is deprecated and is not compatible](https://docs-next.strapi.io/dev-docs/migration/v4-to-v5/breaking-changes/helper-plugin-deprecated) with Strapi 5.) It builds on builds on #133. `useCMEditViewDataManager()` from @strapi/helper-plugin has no direct replacement, but the part used by this plugin is replaced by `useDocument()` from `@strapi/admin/strapi-admin`.

This PR also switches the `<Button />` component usage to `<LinkButton />`.  They look the same, but they allow right clicking to copy links and middle clicking to open in a new tab. If you click them, they behave exactly as the always have; `event.preventDefault()` is called and the regular logic takes over.

### CODE REVIEW

**Comment by mattmilburn** [comment]
Hi @jackbuehner I really appreciate the work you put into this PR! 🎉  I hate to break it to you 💔 that I've already handled a lot of this upgrade before seeing your PR, and I converted everything into TypeScript so it presents quite a conflict with your PR.

However, would you be interested in still rebasing this branch with the latest changes in PR [133](https://github.com/mattmilburn/strapi-plugin-preview-button/pull/133) and migrating the `useCMEditViewDataManager` usage in `EditViewRightLinks` admin component? Also, I wouldn't mind if you wanted to rebase everything and force push the changes to avoid some ugly commit history.

I wonder if that implementation is going to change before the final v5 release but either way, it would be nice because I have not worked with migrating that functionality in a plugin yet 😄 Plus I don't want to steal your Github cred and this would be a good opportunity to be listed as a contributor.

**Comment by jackbuehner** [comment]
No worries! I like the move to TypeScript. I would have posted about contributing on the PR, but I was in the zone on updating everything and just decided to do it.

I'm happy to rebase with your latest changes. It probably can't do it today, but I can definitely do it before the end of the weekend, if that is okay.

They said there shouldn't be too many more big changes now that we are in the release candidate phase, so hopefully this is how it will be. But who knows – they have not yet fully documented the replacements for `useCMEditViewDataManager`.

**Comment by jackbuehner** [comment]
Are you okay with switching `Button` to `LinkButton`? I can also save it for a different PR if you want to keep things separated.

**Comment by mattmilburn** [comment]
> Are you okay with switching `Button` to `LinkButton`? I can also save it for a different PR if you want to keep things separated.

Oh sure, `LinkButton` is fine 👍🏻

File: admin/src/hooks/usePreviewButton.ts [hunk 1]
```diff
@@ -14,17 +15,15 @@ export interface UsePreviewButtonReturn {
 }
 
 const usePreviewButton = (
-  layout: any,
+  uid: UID.ContentType | undefined,
```
Comment (jackbuehner) [inline admin/src/hooks/usePreviewButton.ts:18]: Changed from `layout` to `uid` because we only need `uid`.

`uid` is undefined when the document has not been created yer, which is why we no longer need `isCreating`.

**Comment by jackbuehner** [comment]
Sorry it took a little longer (I got sick). I just did a hard reset to your version of the `strapi5` branch. I then committed on top of that. It should make it so this PR can be cleanly merged into `mattmilburn:strapi5` Let me know if you need me to change anything!

File: admin/src/components/PreviewButton/index.tsx [hunk 2]
```diff
@@ -15,24 +15,25 @@ const PreviewButton = ({ isDraft, openTarget, url }: PreviewButtonProps) => {
 
   const handleClick = useCallback(
     (event: Event) => {
-      if (!url) {
-        event.preventDefault();
+      event.preventDefault();
 
-        return;
-      }
+      if (!url)  return;
```
Comment (mattmilburn) [inline admin/src/components/PreviewButton/index.tsx:20]: Can you include the curly braces here?

File: admin/src/components/PreviewButton/index.tsx [hunk 3]
```diff
@@ -15,24 +15,25 @@ const PreviewButton = ({ isDraft, openTarget, url }: PreviewButtonProps) => {
...
       window.open(url, openTarget);
     },
     [url, openTarget]
   );
 
   return (
-    <Button
+    <LinkButton
+      href={url}
+      target={'_blank'}
```
Comment (mattmilburn) [inline admin/src/components/PreviewButton/index.tsx:30]: Can you simplify this to use double quotes on the attribute? `target="_blank"`

**Comment by mattmilburn** [comment]
@jackbuehner Here are some additional things I found after running `yarn build` to test your updates locally. Do you mind adding these into your PR? Sorry the first item was something I missed before.

1. Add `"styled-components": "^6.1.13"` to both `peerDependencies` and `devDependencies` in `package.json` and run `yarn`.
2. In `ListViewColumnProps` and `AddPreviewColumnProps` can you use `uid: UID.ContentType | undefined` instead of `uid: string`.
3. In `ListViewColumn`, line 38, this hook can be simplified to `usePreviewButton(layout.contentType.uid, data)` to match your update to the hook param.

**Comment by mattmilburn** [comment]
@jackbuehner Also, have you tested your updates in a local Strapi app? When I configure the bare minimum config for a content type, I'm getting this error and I'm not sure what's causing it. Curious if you ran into this or not.

Screenshot 2024-09-18 at 2 30 05 PM

**Comment by jackbuehner** [comment]
The bundle has its own copy of @strapi/admin. This means that the `useStrapiApp` hook cannot find the StrapiApp context because they are separate. This was not a problem earlier in my PR because 

You can verify that the bundled version is being used by adding `console.log('WRONG CONTEXT')` to the `useContext` function inside the `createContext` function in `node_modules/@strapi/admin/dist/admin/Theme-frC82ceE.mjs`:
```
function createContext(rootComponentName, defaultContext) {
  const Context = ContextSelector.createContext(defaultContext);
  const Provider = (props) => {
    const { children, ...context } = props;
    const value = React.useMemo(() => context, Object.values(context));
    return /* @__PURE__ */ jsx(Context.Provider, { value, children });
  };
  const useContext = (consumerName, selector) => ContextSelector.useContextSelector(Context, (ctx) => {
    console.log('WRONG CONTEXT')
    if (ctx)
      return selector(ctx);
    throw new Error(`\`${consumerName}\` must be used within \`${rootComponentName}\``);
  });
  Provider.displayName = rootComponentName + "Provider";
  return [Provider, useContext];
}

const [StrapiAppProvider, useStrapiApp] = createContext("StrapiApp");
```

It was not a problem earlier in my PR because I had used rollup and was not bundling `.mjs` files (which is what `@strapi/admin` was using). https://github.com/mattmilburn/strapi-plugin-preview-button/blob/43f794f46bb32ccd351f37493aee409c2030b1e2/rollup.config.mjs#L24C5-L26C8. I did this out of ignorance; I should have marked `@strapi/admin` as external.

`strapi-plugin` uses rollup behind the scenes. We need to set `@strapi/admin` as external with rollup, but I am not sure how we can do that with `strapi-plugin build`.

The code for build from `strapi-plugin` is at https://github.com/strapi/sdk-plugin/blob/c4702ad5c12fccc83530519a4c95814c09f17bdd/src/cli/commands/plugin/build.ts#L70C5-L75C8. The only options we can provide are the ones available via the command options at the bottom of the file.

Any ideas @mattmilburn

**Comment by mattmilburn** [comment]
@jackbuehner Just an update - I'm reaching out to Strapi support on Discord. This is a popular plugin that even the Strapi team uses internally, so don't worry - we will get this figured out soon enough 🤞🏻

**Comment by jackbuehner** [comment]
Thanks @mattmilburn. My last thought was to use `npx @strapi/sdk-plugin:init` and try to add in bits of the code from this plugin until something breaks (and compare the package versions in package.json). I'll wait until you hear back from support on Discord.

**Comment by jackbuehner** [comment]
Might be related to https://github.com/strapi/strapi/issues/21151

**Comment by danidoff** [comment]
Hey guys. I think they fix it in strapi [5.0.3](https://github.com/strapi/strapi/issues/21151#issuecomment-2399343577)

**Comment by mattmilburn** [comment]
Unfortunately I'm still seeing the same error after updating to v5.0.3.

**Comment by mattmilburn** [comment]
@jackbuehner This is possibly fixed in v5.0.5 🤞🏻

**Comment by SalahAdDin** [comment]
Not able to migrate plugins to version 5 yet!

**Comment by killthekitten** [comment]
@jackbuehner @mattmilburn there was an update from the Strapi team in [this comment](https://github.com/strapi/strapi/issues/21151#issuecomment-2473193358). It appears that the bug is only present when the plugin depends on an RC release of Strapi. Could you try updating the plugin to the latest Strapi? 

In my comment above I might have only bumped the main project's dependency.

**Comment by mattmilburn** [comment]
This is still a problem in v5.4.0. I've created [an issue with Strapi](https://github.com/strapi/strapi/issues/22162) to describe the problem in case anyone is interested.

**Comment by mattmilburn** [comment]
@jackbuehner Problem solved - I was testing this by adding the plugin to a test application's package.json with `file:../<filepath>` and it's actually required to use `yalc` to link a plugin to the Strapi app. See this comment for extra details.

Merging this PR now and I'll cleanup the rest of the plugin details and get it ready to deploy v3 for Strapi 5 🚀 🎉 

> Also might be important to mention that at some point with Strapi v5, they will introduce an **official preview feature** that lets you do a lot more with previewing your editable content within Strapi itself. It looked quite robust when they demo'd it for me, which is exciting. Once that feature is released, I imagine this plugin will be obsolete ❤️

**Review by mattmilburn — APPROVED** [review]

**Comment by jackbuehner** [comment]
> @jackbuehner Problem solved - I was testing this by adding the plugin to a test application's package.json with `file:../<filepath>` and it's actually required to use `yalc` to link a plugin to the Strapi app. See this comment for extra details.

Dang... I was doing the same thing. Really wanted to avoid `yalc`! Thanks for your persistence in troubleshooting the issue you described to the Strapi team in https://github.com/mattmilburn/strapi-plugin-preview-button/pull/146.

**Comment by SalahAdDin** [comment]
> > @jackbuehner Problem solved - I was testing this by adding the plugin to a test application's package.json with `file:../<filepath>` and it's actually required to use `yalc` to link a plugin to the Strapi app. See this comment for extra details.
> 
> Dang... I was doing the same thing. Really wanted to avoid `yalc`! Thanks for your persistence in troubleshooting the issue you described to the Strapi team in #146.

yalc?
